### PR TITLE
e2e: Add e2e for CronFederatedHPA

### PR DESCRIPTION
**Pull request by jwcesign**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Add e2e for CronFederatedHPA.
case 1:Scale FederatedHPA.
case 2:Scale deployment.
case 3:Test suspend rule with CronFederatedHPA

**Which issue(s) this PR fixes**:
Part of https://github.com/karmada-io/karmada/issues/3746

**Special notes for your reviewer**:
none

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Review by chaunceyjiang — COMMENTED** [review]
/assign

File: test/e2e/cronfederatedhpa_test.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+			framework.CreateFederatedHPA(karmadaClient, fhpa)
+		})
+
+		ginkgo.AfterEach(func() {
+			framework.RemoveFederatedHPA(karmadaClient, testNamespace, fhpaName)
+			framework.RemoveCronFederatedHPA(karmadaClient, testNamespace, cronFHPAName)
+		})
+
+		ginkgo.It("Scale FederatedHPA testing", func() {
+			// Wait FederatedHPA to scale deployment's replicas to 1
```
Comment (chaunceyjiang) [inline test/e2e/cronfederatedhpa_test.go:97]: How do you ensure that  the replicas of the deployment can scale-down to 1 when I see that you have set MinReplicas to 1,  MaxReplicas to 10 and AverageUtilization to 80 for FederatedHPA? 


Is there such a situation where `nginx` may consume some CPU, causing the replicas to only scale-down to 2?

Comment (jwcesign) [reply test/e2e/cronfederatedhpa_test.go:97]: We do not request nginx pods, I think scaling down the replicas of nginx is certain.

Comment (chaunceyjiang) [reply test/e2e/cronfederatedhpa_test.go:97]: How about setting  MinReplicas to 1 and MaxReplicas to 1 ?

Comment (jwcesign) [reply test/e2e/cronfederatedhpa_test.go:97]: It's better idea.

File: test/e2e/cronfederatedhpa_test.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+			cronFHPA = helper.NewCronFederatedHPAWithScalingDeployment(testNamespace, cronFHPAName, deploymentName, rule)
+		})
+
+		ginkgo.AfterEach(func() {
+			framework.RemoveCronFederatedHPA(karmadaClient, testNamespace, cronFHPAName)
+		})
+
+		ginkgo.It("Scale Deployment testing", func() {
+			// The init replicas of deployment should be 3
+			framework.WaitDeploymentReplicasFitWith(framework.ClusterNames(), testNamespace, deploymentName, 3)
```
Comment (chaunceyjiang) [inline test/e2e/cronfederatedhpa_test.go:121]:
```suggestion
			framework.WaitDeploymentReplicasFitWith(framework.ClusterNames(), testNamespace, deploymentName, int(*deployment.Spec.Replicas))
```

File: test/e2e/cronfederatedhpa_test.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+		ginkgo.It("Test suspend rule with CronFederatedHPA", func() {
+			// The init replicas of deployment should be 3
+			framework.WaitDeploymentReplicasFitWith(framework.ClusterNames(), testNamespace, deploymentName, 3)
+
+			// Create CronFederatedHPA to scale Deployment
+			framework.CreateCronFederatedHPA(karmadaClient, cronFHPA)
+
+			// Wait 3 minutes and check whether the replicas changed
+			time.Sleep(time.Minute*1 + time.Second*30)
```
Comment (chaunceyjiang) [inline test/e2e/cronfederatedhpa_test.go:150]:
```suggestion
			time.Sleep(pollTimeout / 2)
```

Comment (jwcesign) [reply test/e2e/cronfederatedhpa_test.go:150]: Why? The CronFederatedHPA is supposed to execute every minute. I will wait for 1:30 and check if the replicas have changed.

Comment (chaunceyjiang) [reply test/e2e/cronfederatedhpa_test.go:150]: I understand, but `(time.Minute*1 + time.Second*30)` is a magic number, so I suggest using pollTimeout instead.

Comment (RainbowMango) [reply test/e2e/cronfederatedhpa_test.go:150]: By the way,
wait `3 minutes`?  Waiting time seems too long which will seriously slow down the test progress.

Comment (jwcesign) [reply test/e2e/cronfederatedhpa_test.go:150]: We should wait for this CronFederatredHPA rule, and check whether it's suspended, I didn't find other way to solve this.

Comment (jwcesign) [reply test/e2e/cronfederatedhpa_test.go:155]: `pollTimeout = 420 * time.Second` is too long, I add some comments to make it clearly:
```
// */1 * * * * means the rule will be triggered every 1 minute
// So wait for 1m30s and check whether the replicas changed and whether the suspend field works
```

File: test/e2e/cronfederatedhpa_test.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+/*
+CronFederatedHPA focus on scaling FederatedHPA or other resource with scale subresource (e.g. Deployment, StatefulSet).
+Test Case Overview:
+
+	case 1:
+		Scale FederatedHPA.
+	case 2:
+		Scale deployment.
+	case 3:
+		Test suspend rule with CronFederatedHPA
```
Comment (chaunceyjiang) [inline test/e2e/cronfederatedhpa_test.go:39]: I think we need to add another case:  **modify** the existing rule("suspend=false") to "suspend=true".

**Comment by jwcesign** [comment]
/cc @chaunceyjiang @XiShanYongYe-Chang 
PTAL

**Review by chaunceyjiang — COMMENTED** [review]
LGTM

**Comment by jwcesign** [comment]
/cc @ikaven1024
Please take a look at the script updates.

File: test/e2e/cronfederatedhpa_test.go (hunk 3 above)
Comment (RainbowMango) [reply test/e2e/cronfederatedhpa_test.go:195]: @chaunceyjiang What do you think?
I didn't look into it.

Comment (chaunceyjiang) [reply test/e2e/cronfederatedhpa_test.go:195]: I also didn't find other way to solve this.

**Review by chaunceyjiang — COMMENTED** [review]
/lgtm

File: test/e2e/cronfederatedhpa_test.go (hunk 3 above)
Comment (XiShanYongYe-Chang) [reply test/e2e/cronfederatedhpa_test.go:195]: Wait until can be a slightly more elegant way to modify it. Although fundamentally still need to wait for a little over a minute.

**Pull request merged**